### PR TITLE
Insecure Authorization Restriction KB Entry 

### PR DESCRIPTION
**Pull request by ohachimOs**
4xx bypass KB entry.

### CODE REVIEW

File: WEB_SERVICE/WEB/_HIGH/4XX_BYPASS/description.md [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        ),
+      ),
+    );
+  }
+
+  void fetchResource(String url) {
+    // Code to make HTTP request to the provided URL
+  }
+}
+```
```
Comment (rabsondev) [inline WEB_SERVICE/WEB/_HIGH/4XX_BYPASS/description.md:61]: For the code samples, they should be added with the custom tabs syntax (look at [this example](https://raw.githubusercontent.com/Ostorlab/KB/dcad0201634357222a1885d791f15ac572ff1d55/MOBILE_CLIENT/COMMON/_MEDIUM/OAUTH_SCHEME_ACCOUNT_TAKEOVER/description.md))
image

File: WEB_SERVICE/WEB/_HIGH/4XX_BYPASS/description.md [hunk 2]
```diff
@@ -0,0 +1,19 @@
+
+# 4XX-Bypass
+
+The 4xx-bypass exploits refer to HTTP request fuzzing techniques that allow attackers to bypass server-side restrictions, accessing unauthorized resources or performing malicious actions on a web application. This can lead to unauthorized data access, privilege escalation, and other security risks.
+
+=== "Python example"
```
Comment (deadly-panda) [inline WEB_SERVICE/WEB/_HIGH/4XX_BYPASS/description.md:6]: this should show the language.. please check the other entries for the structure.

Comment (deadly-panda) [inline WEB_SERVICE/WEB/_HIGH/4XX_BYPASS/description.md:6]:
```suggestion
=== "Python"
```

File: WEB_SERVICE/WEB/_HIGH/4XX_BYPASS/description.md [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+  ```python
+    import requests
+
+    response = requests.get("http://www.some-url.com/unauthorized_path")
+
+    '''if we have some unauthorized path that gets us a 403 code,
+    we can try something like adding a query parameter to see if we can
+    trick the server by exploiting some mistake.'''
```
Comment (deadly-panda) [inline WEB_SERVICE/WEB/_HIGH/4XX_BYPASS/description.md:14]: would be nice to give an example on how the backend would handle this to show the vuln.

File: WEB_SERVICE/WEB/_HIGH/4XX_BYPASS/meta.json [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+    "short_description": "4xx-bypass vulnerability allows an attacker to bypass access controls by manipulating HTTP requests to gain access to restricted resources or perform actions they are not supposed to.",
+    "references": {
+        "wikipedia": "https://en.wikipedia.org/wiki/List_of_HTTP_status_codes#4xx_client_errors",
+        "Hacktricks": "https://book.hacktricks.xyz/network-services-pentesting/pentesting-web/403-and-401-bypasses",
+        "cheatsheetseries.owasp": "https://cheatsheetseries.owasp.org/assets/REST_Security_Cheat_Sheet_Bypassing_VBAAC_with_HTTP_Verb_Tampering.pdf"
+    },
+    "title": "4xx-bypass Vulnerability",
+    "privacy_issue": true,
+    "security_issue": true
+}
```
Comment (deadly-panda) [inline WEB_SERVICE/WEB/_HIGH/4XX_BYPASS/meta.json:12]: I think you are missing the standars here..

File: WEB_SERVICE/WEB/_HIGH/4XX_BYPASS/description.md (hunk 3 above)
Comment (ohachimOs) [reply WEB_SERVICE/WEB/_HIGH/4XX_BYPASS/description.md:14]: @deadly-panda can you please check the example I provided

File: WEB_SERVICE/WEB/_HIGH/4XX_BYPASS/recommendation.md [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+      if request.method == 'GET':
+          return jsonify({"message": "This is a GET request"})
+      elif request.method == 'POST':
+          data = request.json
+          return jsonify({"message": "This is a POST request", "data": data})
+   ```
+   ```python
+   # only use headers you expect
+   @app.route('/limiting_header_values')
+  def limiting_header_values():
```
Comment (ErebusZ) [inline WEB_SERVICE/WEB/_HIGH/4XX_BYPASS/recommendation.md:26]: I think one is enough per language

Comment (ohachimOs) [reply WEB_SERVICE/WEB/_HIGH/4XX_BYPASS/recommendation.md:26]: Okey

File: WEB_SERVICE/WEB/_HIGH/4XX_BYPASS/meta.json [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    "privacy_issue": true,
+    "security_issue": true,
+    "categories": {
+        "OWASP_ASVS_L1": [
+            "V4_1_1",
+            "V4_1_2",
+            "V4_1_3",
+            "V4_1_5",
+            "V5_1_1",
+            "V5_1_2",
```
Comment (ErebusZ) [inline WEB_SERVICE/WEB/_HIGH/4XX_BYPASS/meta.json:19]: you need to use the standards keys defined in RE.

File: WEB_SERVICE/WEB/_HIGH/4XX_BYPASS/description.md [hunk 7]
```diff
@@ -0,0 +1,32 @@
+
+# 4XX-Bypass
+
+The 4xx-bypass exploits refer to HTTP request fuzzing techniques that allow attackers to bypass server-side restrictions, accessing unauthorized resources or performing malicious actions on a web application. This can lead to unauthorized data access, privilege escalation, and other security risks.
```
Comment (oussamaessaji) [inline WEB_SERVICE/WEB/_HIGH/4XX_BYPASS/description.md:4]: Needs more details (other security risks) is not clear???

File: WEB_SERVICE/WEB/_HIGH/4XX_BYPASS/description.md [hunk 8]
```diff
@@ -0,0 +1,32 @@
+
+# 4XX-Bypass
```
Comment (3asm) [inline WEB_SERVICE/WEB/_HIGH/4XX_BYPASS/description.md:2]: Remove, title is in the meta.json.

File: WEB_SERVICE/WEB/_HIGH/4XX_BYPASS/description.md (hunk 7 above)
Comment (3asm) [inline WEB_SERVICE/WEB/_HIGH/4XX_BYPASS/description.md:4]: Nah nah nah. There is no such thing as a 4xx bypass. Call it Insecure Authorization Restriction and change all the entries.

File: WEB_SERVICE/WEB/_HIGH/4XX_BYPASS/description.md [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+The 4xx-bypass exploits refer to HTTP request fuzzing techniques that allow attackers to bypass server-side restrictions, accessing unauthorized resources or performing malicious actions on a web application. This can lead to unauthorized data access, privilege escalation, and other security risks.
+
+=== "Python"
+  ```python
```
Comment (3asm) [inline WEB_SERVICE/WEB/_HIGH/4XX_BYPASS/description.md:7]: Add other languages.

File: WEB_SERVICE/WEB/_HIGH/4XX_BYPASS/description.md [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+    we can try something like adding a query parameter to see if we can
+    trick the server by exploiting some mistake.'''
+
+    response = requests.get("http://www.some-url.com/unauthorized_path?debug=true")
+
+    '''Might get us the resource we want'''
+  ```
+
+=== "Python"
+  ```python
```
Comment (3asm) [inline WEB_SERVICE/WEB/_HIGH/4XX_BYPASS/description.md:22]: Both Python, why?

File: WEB_SERVICE/WEB/_HIGH/4XX_BYPASS/recommendation.md [hunk 11]
```diff
@@ -0,0 +1,93 @@
+
+# 4XX-Bypass
+
```
Comment (3asm) [inline WEB_SERVICE/WEB/_HIGH/4XX_BYPASS/recommendation.md:3]: Remove title.

**Review by 3asm — CHANGES_REQUESTED** [review]
Needs way more work.

File: WEB_SERVICE/WEB/_HIGH/4XX_BYPASS/meta.json [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+    "short_description": "4xx-bypass vulnerability allows an attacker to bypass access controls by manipulating HTTP requests to gain access to restricted resources or perform actions they are not supposed to.",
+    "references": {
+        "wikipedia": "https://en.wikipedia.org/wiki/List_of_HTTP_status_codes#4xx_client_errors",
+        "Hacktricks": "https://book.hacktricks.xyz/network-services-pentesting/pentesting-web/403-and-401-bypasses",
+        "cheatsheetseries.owasp": "https://cheatsheetseries.owasp.org/assets/REST_Security_Cheat_Sheet_Bypassing_VBAAC_with_HTTP_Verb_Tampering.pdf"
+    },
+    "title": "4xx-bypass Vulnerability",
+    "privacy_issue": true,
+    "security_issue": true,
+    "categories": {
```
Comment (elyousfi5) [inline WEB_SERVICE/WEB/_HIGH/4XX_BYPASS/meta.json:12]: are you sure there are no PCI standards to add ?

File: WEB_SERVICE/WEB/_HIGH/4XX_BYPASS/description.md (hunk 7 above)
Comment (ohachimOs) [reply WEB_SERVICE/WEB/_HIGH/4XX_BYPASS/description.md:4]: Okey!

File: WEB_SERVICE/WEB/_HIGH/4XX_BYPASS/description.md (hunk 10 above)
Comment (ohachimOs) [reply WEB_SERVICE/WEB/_HIGH/4XX_BYPASS/description.md:22]: I removed the doubled code block ticks

File: WEB_SERVICE/WEB/_HIGH/4XX_BYPASS/meta.json [hunk 13]
```diff
@@ -0,0 +1,47 @@
+{
+    "risk_rating": "high",
+    "short_description": "Insecure Authorization Restriction vulnerability allows an attacker to bypass access controls by manipulating HTTP requests to gain access to restricted resources or perform actions they are not supposed to.",
+    "references": {
+        "wikipedia": "https://en.wikipedia.org/wiki/List_of_HTTP_status_codes#4xx_client_errors",
+        "Hacktricks": "https://book.hacktricks.xyz/network-services-pentesting/pentesting-web/403-and-401-bypasses",
+        "cheatsheetseries.owasp": "https://cheatsheetseries.owasp.org/assets/REST_Security_Cheat_Sheet_Bypassing_VBAAC_with_HTTP_Verb_Tampering.pdf"
+    },
+    "title": "Insecure Authorization Restriction Vulnerability",
```
Comment (elyousfi5) [inline WEB_SERVICE/WEB/_HIGH/4XX_BYPASS/meta.json:9]: Can you please change the folder of the KB to something like

`WEB_SERVICE/WEB/_HIGH/INSECURE_AUTHORIZATION_RESTRICTION`

File: WEB_SERVICE/WEB/_HIGH/4XX_BYPASS/meta.json (hunk 12 above)
Comment (ohachimOs) [reply WEB_SERVICE/WEB/_HIGH/4XX_BYPASS/meta.json]: One PCI standard that looked like it fits!

File: WEB_SERVICE/WEB/_HIGH/4XX_BYPASS/description.md (hunk 10 above)
Comment (ohachimOs) [reply WEB_SERVICE/WEB/_HIGH/4XX_BYPASS/description.md]: I added other language for the backend/server part

File: WEB_SERVICE/WEB/_HIGH/INSECURE_AUTHORIZATION_RESTRICTION/meta.json [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+            "V4_1_3",
+            "V4_1_5",
+            "V5_1_1",
+            "V5_1_2",
+            "V5_1_3",
+            "V5_1_4",
+            "V5_2_2"
+        ],
+        "PCI_STANDARDS": [
+            "REQ_2_2 "
```
Comment (elyousfi5) [inline WEB_SERVICE/WEB/_HIGH/INSECURE_AUTHORIZATION_RESTRICTION/meta.json:47]:
```suggestion
            "REQ_2_2"
```

File: WEB_SERVICE/WEB/_HIGH/4XX_BYPASS/meta.json (hunk 13 above)
Comment (ohachimOs) [reply WEB_SERVICE/WEB/_HIGH/4XX_BYPASS/meta.json]: Done!

File: WEB_SERVICE/WEB/_HIGH/INSECURE_AUTHORIZATION_RESTRICTION/description.md [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+
+        if (isset($queryParams['debug']) && $queryParams['debug'] === 'true') {
+            // Logic that would usually require more rights
+            echo 'Debug mode enabled: Performing privileged actions.';
+        } else {
+            echo 'Normal mode: Limited actions.';
+        }
+    }
+    ?>
+   ```
```
Comment (3asm) [inline WEB_SERVICE/WEB/_HIGH/INSECURE_AUTHORIZATION_RESTRICTION/description.md:66]:
```suggestion
Insecure Authorization Restriction refers to weaknesses in server-side restrictions that can be exploited through HTTP request manipulation techniques. This vulnerability allows attackers to bypass access controls, leading to unauthorized access to resources, privilege escalation, and allowing unauthorized users to retrieve, create, update, or delete sensitive data.

Targeting Insecure Authorization Restrictions on a web server, for example, would involve these techniques:

   * **HTTP Request Method fuzzing**: Testing invalid, malformed, or unexpected HTTP methods on the server.
   * **HTTP Request Path fuzzing**: Manipulating the HTTP request path by deforming it, adding to it, or subtracting from it.
   * **HTTP Request Query Paramete**r fuzzing: Adding, removing, or changing the original query parameters.
   * **HTTP Request Header fuzzing**: Adding familiar headers, removing headers, or adding proxy headers.

All these techniques target defects, vulnerabilities, or mistakes in a server's logic.

=== "Python"
  ```python
    import requests

    response = requests.get("http://www.some-url.com/unauthorized_path")

    '''if we have some unauthorized path that gets us a 403 code,
    we can try something like adding a query parameter like "debug=true" to see if we can
    trick the server by exploiting some mistake.'''

    response = requests.get("http://www.some-url.com/unauthorized_path?debug=true")

    '''Might get us the resource we want''
  ```
=== "JavaScript"
   ```javascript

      const express = require('express');
      const app = express();

      app.get('/vulnerable_path', (req, res) => {
        const queryParams = req.query;

        if (queryParams.debug && queryParams.debug === 'true') {
          // Logic that would usually require more rights
          res.send('Debug mode enabled: Performing privileged actions.');
        } else {
          res.send('Normal mode: Limited actions.');
        }
      });

      app.listen(3000, () => {
          console.log('Server is running on port 3000');
      });
   ```
=== "PHP"
   ```php
    <?php
    // Simple example of a PHP backend that forgot a debug conditional for debugging purposes

    if ($_SERVER['REQUEST_METHOD'] === 'GET' || $_SERVER['REQUEST_METHOD'] === 'POST') {
        $queryParams = $_GET; // Using $_GET to retrieve query parameters from the URL

        if (isset($queryParams['debug']) && $queryParams['debug'] === 'true') {
            // Logic that would usually require more rights
            echo 'Debug mode enabled: Performing privileged actions.';
        } else {
            echo 'Normal mode: Limited actions.';
        }
    }
    ?>
   ```
```

File: WEB_SERVICE/WEB/_HIGH/INSECURE_AUTHORIZATION_RESTRICTION/recommendation.md [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+                    
+                 if (expectedHeader == null) {
+                     return ResponseEntity.badRequest().body("Expected-Header is missing");
+                 }
+
+                 // Logic depending on the expected header
+                 return ResponseEntity.ok("Received expected header: " + expectedHeader);
+             }
+         }
+   ```
```
Comment (3asm) [inline WEB_SERVICE/WEB/_HIGH/INSECURE_AUTHORIZATION_RESTRICTION/recommendation.md:116]:
```suggestion
Organizations should implement proper access controls and enforce strict validation of HTTP requests to mitigate the risk of insecure authorization restriction HTTP vulnerabilities. This involves having robust server-side logic to manage rules over the HTTP methods, headers, query parameters, and paths received.

Here are some recommendations:

  * **Limit HTTP Methods**: Restrict which HTTP methods can access each of your views/resources.
  * **Sanitize Query Parameters**: Sanitize each request's query parameters and ensure only a limited set of parameters can affect the server's logic.
  * **Limit Headers**: Restrict which headers can affect your code. Use strict rules on what header/method combinations can make changes on the server side, and ensure your logic does not rely on header values that can be easily found on the internet (like Google's User-Agent).
  * **Robust Path Parsing**: Implement robust path parsing with strict rules and reject requests that do not conform to your standards.

=== "Python"
   ```python
   # allow only GET and POST methods for this route
  @app.route('/limiting_method_usage', methods=['GET', 'POST'])
  def limiting_method_usage():
      if request.method == 'GET':
          return jsonify({"message": "This is a GET request"})
      elif request.method == 'POST':
          data = request.json
          return jsonify({"message": "This is a POST request", "data": data})
   # only use headers you expect
   @app.route('/limiting_header_values')
  def limiting_header_values():
      expected_header = request.headers.get('Expected-Header')
      ### logic depending on the expected header
   ```


=== "PHP"
   ```php
    <?php
    // Assuming you have a framework like Slim or a similar one installed
    require 'vendor/autoload.php';

    $app = new \Slim\App;

    $app->get('/sanitize/{url_path:.*}', function ($request, $response, $args) {
        $urlPath = $args['url_path'];
        $sanitizedPath = sanitizePath($urlPath);

        if ($sanitizedPath === null) {
            return $response->withStatus(400)->withJson(["description" => "Invalid URL path"]);
        }
    
        return $response->withJson(["message" => "Path is sanitized", "sanitized_path" => $sanitizedPath]);
    });

    function sanitizePath($path) {
        // Implement your sanitization logic here
        if (preg_match('/^[a-zA-Z0-9]+$/', $path)) {
            return $path;
        }
        return null;
    }

    $app->run();
    ?>
   ```


=== "JavaScript"
   ```javascript

        const express = require('express');
        const app = express();

        app.get('/sanitize/:url_path(*)', (req, res) => {
            const urlPath = req.params.url_path;
            const sanitizedPath = sanitizePath(urlPath);

            if (sanitizedPath === null) {
                return res.status(400).json({ description: "Invalid URL path" });
            }

            res.json({ message: "Path is sanitized", sanitized_path: sanitizedPath });
        });

        function sanitizePath(path) {
            // Implement your sanitization logic here
            if (/^[a-zA-Z0-9]+$/.test(path)) {
                return path;
            }
            return null;
        }

        const port = 3000;
        app.listen(port, () => {
            console.log(`Server is running on port ${port}`);
        });
   ```

=== "Java"
   ```java

         import org.springframework.http.ResponseEntity;
         import org.springframework.web.bind.annotation.GetMapping;
         import org.springframework.web.bind.annotation.RequestHeader;
         import org.springframework.web.bind.annotation.RequestMapping;
         import org.springframework.web.bind.annotation.RestController;

         @RestController
         @RequestMapping("/api")
         public class HeaderController {

             @GetMapping("/limiting_header_values")
             public ResponseEntity<String> limitingHeaderValues(
                     @RequestHeader(value = "Expected-Header", required = false) String expectedHeader) {
                    
                 if (expectedHeader == null) {
                     return ResponseEntity.badRequest().body("Expected-Header is missing");
                 }

                 // Logic depending on the expected header
                 return ResponseEntity.ok("Received expected header: " + expectedHeader);
             }
         }
   ```
```

**Review by 3asm — APPROVED** [review]

**Pull request merged**